### PR TITLE
Replace raw streams with rmm::cuda_stream_view (part 2)

### LINKED ISSUE #6706: [BUG] cudf::to_arrow is not properly synchronized
`cudf::to_arrow` copies data from device to host on a stream and never synchronizes that stream before making it available to the Arrow library, which does not use the same stream. Therefore I believe there is a potential race condition. 

This is an example of one of the functions in `to_arrow.cpp` that copies data asynchronously to the host.

https://github.com/rapidsai/cudf/blob/3d44ed58ca95f4538b059fe4ac7f9cf17b6b4867/cpp/src/interop/to_arrow.cpp#L39-L58

The `to_arrow` function ultimately calls this function and others like it, but never synchronizes before returning. Therefore an application could use this function to get an Arrow array/buffer, and then access that array on the host without synchronizing the stream.

I believe currently pageable host memory is used, which would cause `cudaMemcpyAsync` to return only once the copy is completed. But if this is switched to pinned host memory, it will be asynchronous, and we will have a race condition.

### PR DESCRIPTION
**Pull request by harrism**
Converting libcudf to use rmm::cuda_stream_view will require a LOT of changes, so I'm splitting it into multiple PRs to ease reviewing. This is the second PR in the series. This series of PRs will

 - Replace usage of cudaStream_t with rmm::cuda_stream_view
 - Replace usage of 0 or nullptr as a stream identifier with rmm::cuda_stream_default
 - Ensure all APIs always order the stream parameter before the memory resource parameter. #5119

Contributes to #6645 and #5119

Depends on #6646 so this PR will look much bigger until that one is merged.

Also fixes #6706  (to_arrow and to_dlpack are not properly synchronized).

This second PR converts:

 - table.hpp (and source / dependencies)
 - column_factories.hpp (and source / dependencies)
 - headers in include/detail/aggregation (and source / dependencies)
 - include/detail/groupby/sort_helper.hpp (and source / dependencies)
 - include/detail/utilities/cuda.cuh (and dependencies)
 - binary ops
 - concatenate
 - copy_if
 - copy_range
 - fill
 - gather
 - get_value
 - hash groupby
 - quantiles
 - reductions
 - repeat
 - replace
 - reshape
 - round
 - scatter
 - search
 - sequence
 - sorting
 - stream compaction

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Review by ajschmidt8 — APPROVED** [review]

**Comment by harrism** [comment]
rerun tests

**Comment by harrism** [comment]
This PR no longer requires cmake or java codeowners approval.

**Review by revans2 — APPROVED** [review]
A little hard to do the review with 1000+ lines of auto-reformatting on the java side, but I got through it and I guess it is better to pull off the bandaid.

**Review by kkraus14 — APPROVED** [review]
Cmake / conda approval

**Comment by harrism** [comment]
> A little hard to do the review with 1000+ lines of auto-reformatting on the java side, but I got through it and I guess it is better to pull off the bandaid.

So sorry I'm so used to clang-format now I didn't even notice that...  Maybe we should turn on clang-format for the java directory in @codereport 's clang-format upgrade PR?

File: cpp/src/copying/scatter.cu
Comment (devavret) [inline cpp/src/copying/scatter.cu:188]: mr is a default param now.

File: cpp/src/dictionary/set_keys.cu
Comment (devavret) [inline cpp/src/dictionary/set_keys.cu:159]: can remove mr here

File: cpp/include/cudf/detail/concatenate.cuh
Comment (harrism) [reply cpp/include/cudf/detail/concatenate.cuh:41]: I removed a lot of default params for streams where they are not needed. IMO default params on streams lead people to forget to use them in places where they should be using them. I think in a future PR we may remove all default stream parameters.

Comment (harrism) [reply cpp/include/cudf/detail/concatenate.cuh:41]: FWIW, this function never had one anyway.

File: cpp/include/cudf/detail/quantiles.hpp
Comment (harrism) [reply cpp/include/cudf/detail/quantiles.hpp:26]: This is just what clangd does when a line is too long. If I add the `*`, it will more than likely screw up doxygen finding the doc to copy, right?

Comment (harrism) [reply cpp/include/cudf/detail/quantiles.hpp:26]: Aha, looking at doxygen docs, it appears you only need the parameter list if the function whose docs you are copying is overloaded? Will try without it.

File: cpp/src/copying/scatter.cu
Comment (harrism) [reply cpp/src/copying/scatter.cu:188]: I don't understand this comment. `dictionary::detail::get_index()` has no default parameters.

File: cpp/src/dictionary/set_keys.cu
Comment (harrism) [reply cpp/src/dictionary/set_keys.cu:159]: We can, or we can leave it explicit to make it clear we're allocating a temporary...

Comment (devavret) [reply cpp/src/dictionary/set_keys.cu:159]: That wouldn't be consistent with what we do everywhere else. If `mr` is passed along then the returned allocation is not temporary otherwise it is.

File: cpp/src/copying/scatter.cu
Comment (devavret) [reply cpp/src/copying/scatter.cu:188]: https://github.com/rapidsai/cudf/pull/6648/files?file-filters%5B%5D=.hpp#diff-da300662b4bc48b59f81fbba20725eacf62dbabb0a0c671e2fa8b05aeefca775R32-R36
 Similar suggestion as https://github.com/rapidsai/cudf/pull/6648#discussion_r523259260 of not needing mr.

File: cpp/src/sort/sort.cu
Comment (cwharris) [inline cpp/src/sort/sort.cu:17]:
```suggestion
<include "rmm/cuda_stream_view.hpp>
```

File: cpp/src/stream_compaction/drop_nulls.cu
Comment (cwharris) [inline cpp/src/stream_compaction/drop_nulls.cu:24]:
```suggestion
<include "rmm/cuda_stream_view.hpp>
```

File: cpp/src/transform/transform.cpp
Comment (cwharris) [inline cpp/src/transform/transform.cpp:29]:
```suggestion
#include <rmm/cuda_stream_view.hpp>
```

File: cpp/src/unary/nan_ops.cu
Comment (cwharris) [inline cpp/src/unary/nan_ops.cu:24]:
```suggestion
#include <rmm/cuda_stream_view.hpp>
```

File: cpp/src/unary/unary_ops.cuh
Comment (cwharris) [inline cpp/src/unary/unary_ops.cuh:26]:
```suggestion
#include <rmm/cuda_stream_view.hpp>
```

File: java/src/main/native/src/map_lookup.cu
Comment (cwharris) [inline java/src/main/native/src/map_lookup.cu:30]:
```suggestion
#include <rmm/cuda_stream_view.hpp>
```

**Review by cwharris — APPROVED** [review]
Some comments, but otherwise LGTM! :fire:

File: cpp/src/copying/scatter.cu
Comment (harrism) [reply cpp/src/copying/scatter.cu]: OK, you convinced me. I changed your two suggestions and all the others I could find to not call `get_current_device_resource` when that is the default.

File: cpp/src/dictionary/detail/concatenate.cu
Comment (harrism) [reply cpp/src/dictionary/detail/concatenate.cu:28]: Both `cuda_stream_view` and `device_uvector` are used multiple times in this file.

File: cpp/src/interop/dlpack.cpp
Comment (harrism) [reply cpp/src/interop/dlpack.cpp:24]: I can't find-replace it because in all instances the headers also need to be reordered / regrouped.

Comment (harrism) [reply cpp/src/interop/dlpack.cpp:24]: Most of these are fixed in the part-3 PR, FWIW.

Comment (harrism) [reply cpp/src/interop/dlpack.cpp:24]: Fixed.

File: cpp/src/dictionary/set_keys.cu
Comment (harrism) [reply cpp/src/dictionary/set_keys.cu]: Well, not passing it and passing `get_current_device_resource` are equivalent -- both do the right thing for temporaries, and the latter is more explicit at the point of call. But I have changed all instances, including the two pointed out.

**Comment by harrism** [comment]
OK, I have now addressed all review comments @devavret @cwharris . Thanks for catching lots of easily missed things!

**Review by devavret — APPROVED** [review]

**Pull request merged**
